src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1517,18 +1517,33 @@
       containing_class text,
       corrective_change boolean)")
 
+  ;; null values are not considered equal in postgresql indexes,
+  ;; therefore the existing unique constraint did not function
+  ;; as intended. this will replace the unique constraint with
+  ;; a primary key that is a hash of the four columns used to
+  ;; consider uniqueness.
+
+  ;; since the unique constraint could allow duplicates to exist
+  ;; in the table, we have to keep a running set of the hashes
+  ;; we've encountered during the migration to avoid inserting
+  ;; duplicates into the new table.
+
   (jdbc/call-with-query-rows
    ["select
        report_id, certname_id, status, \"timestamp\", resource_type, resource_title, property,
        new_value, old_value, message, file, line, containment_path, containing_class, corrective_change
      from resource_events"]
    (fn [rows]
-     (doseq [batch (partition-all 500 rows)]
-       (jdbc/insert-multi! "resource_events_transform"
-                           (map #(-> %
-                                     (assoc :event_hash (->> (hash/resource-event-identity-pkey %)
-                                                             (sutils/munge-hash-for-storage))))
-                                batch)))))
+     (reduce (fn [hashes-seen row]
+               (conj hashes-seen
+                     (let [hash-str (hash/resource-event-identity-pkey row)]
+                       (when-not (hashes-seen hash-str)
+                         (jdbc/insert-multi! "resource_events_transform"
```

Comment:
Does this mean we're not batching the inserts here anymore? Just want to make sure I know what it's doing.

Reply:
no, we're not - but it happens in a single transaction, so it evens out.

Reply:
Yeah, perf remained fairly consistent between the two. Ordinarily I would batch, but that definitely makes the migration code here more complex.

I'll keep `insert!` vs `insert-multi!` in my head for next time. I just left it as `insert-multi!` because it was there prior to this change.